chain-vote/src/tally.rs

```diff
@@ -223,11 +223,17 @@ impl ValidatedTally {
     pub fn decrypt_tally(
         &self,
         max_votes: u64,
-        table: &TallyOptimizationTable,
+        table: &Option<TallyOptimizationTable>,
```

Comment:
```suggestion
        table: Option<&TallyOptimizationTable>,
```

seems more sensible

Reply:
@Zeegomo any thoughts on this?

Reply:
Thought in this?

> Another way to go there is to cover this special case with an enum and protect ourselves from such an error on the type system level. Something like max_votes_with_optimization_table: `Option<(NonZeroU64, &TallyOptimizationTable)>`

Reply:
I've made an enum that looks like:
```rust
enum TallyDecryptStrategy<'a> {
  WithVotes(&'a BabyStepTable, NonZeroU64),
  WithoutVotes,
}
```
This seems to deal with the "ability to misuse the API", as well as opting into null pointer optimization. I'm still not 100% familiar with the code, so it's possible I've chosen the wrong names for things though. API-wise, does this convey the intent better?

Reply:
Right, it solves the misuse problem, but I would argue that it does not make a lot of sense to call the function with  `WithoutVotes` variant? It's already clear that the function has nothing do to in that case and the caller already knows the result.

Reply:
Currently it calls the private `decrypt` function, which in turn accesses private fields, so it seems like there isn't a way to work out the length of the `Tally` vector from outside this module. But I'm not super clear on the model yet so maybe we can get it from something in `vote.rs`?